hardware/can/can.c

```diff
@@ -0,0 +1,372 @@
+	}
+	return count;
+}
+
+/*---------------------------------------------------------------------
+ Gesamtzahl empfangender CAN Messages
+ * --------------------------------------------------------------------*/
+uint32_t can_get_counter_messages_received()
+{
+	return can_messages_received;
+}
+
+/*---------------------------------------------------------------------
+ Gesamtzahl gesendeter CAN Messages
+ * --------------------------------------------------------------------*/
+uint32_t can_get_counter_messages_send()
+{
+	return can_messages_send;
+}
+
+/*---------------------------------------------------------------------
+ Max Anzahl im FIFO TX
+ * --------------------------------------------------------------------*/
+uint8_t can_get_counter_fifo_max_tx()
+{
+	return fifo_max_tx;
+}
+
+/*---------------------------------------------------------------------
+ Max Anzahl im FIFO RX
+ * --------------------------------------------------------------------*/
+uint8_t can_get_counter_fifo_max_rx()
+{
+	return fifo_max_rx;
+}
+
+/*---------------------------------------------------------------------
+ Counter reset
+ * --------------------------------------------------------------------*/
+void can_reset_counter()
+{
+	can_messages_received = 0;
+	can_messages_send = 0;
+	fifo_max_tx = 0;
+	fifo_max_rx = 0;
+}
+
+/*---------------------------------------------------------------------
+ Sendet CAN-Message
+ * --------------------------------------------------------------------*/
+bool can_send_message(can_t *message)
+{
+	uint8_t ret;
+
+	ATOMIC_BLOCK(ATOMIC_RESTORESTATE)
+	{
+		ret = fifo_push(&fifo_tx, message);
+
+		if (fifo_tx.count > fifo_max_tx)
+		{
+			fifo_max_tx = fifo_tx.count;
+		}
+	}
+	return ret;
+}
+
+/*---------------------------------------------------------------------
+ Versendet CAN Message: Übertragung vom FIFO in den MCP2515
+ Diese Funktion muss regelmäßig aufgerufen werden
+ * --------------------------------------------------------------------*/
+void can_send_tx_fifo()
+{
+	can_t message;
+	uint8_t tx_buffer = 0;
+	uint8_t status;
+
+	ATOMIC_BLOCK(ATOMIC_RESTORESTATE)
+	{
+		// Message(s) im FIFO?
+		while (fifo_check(&fifo_tx))
+		{
+			// Zustand der drei TX Buffer
+			status = mcp2515_get_free_buffer();
+
+			// TX-Buffer 1 frei
+			if (bit_is_clear(status, 2))
+				tx_buffer = 1;
+			// TX-Buffer 2 frei
+			else if (bit_is_clear(status, 4))
+				tx_buffer = 2;
+			// TX-Buffer 3 frei
+			else if (bit_is_clear(status, 6))
+				tx_buffer = 3;
+
+			if (tx_buffer > 0)
+			{
+				fifo_pull_nowait(&fifo_tx, &message);
+				mcp2515_send_message(&message, tx_buffer);
+				can_messages_send++;
+				tx_buffer = 0;
+			}
+			else
+			{
+				// Alle TX-Buffer voll
+				break;
+			}
+		}
+	}
+}
+
+/*---------------------------------------------------------------------
+ Print CAN Message
+ * --------------------------------------------------------------------*/
+void can_print_message(const can_t *message)
+{
```

Comment:
Please do not mix test or example code with the functional code, at least not without an #ifdef.